### DISCUSSION TITLE
Duplicate network requests in network tab

### DISCUSSION THREAD

**Discussion by nandorojo** [urql-graphql/urql#2576] [Q&A]
I've noticed that my `useQuery` calls are sending the same request twice when I look in the network tab. Has anyone else seen this? The requests have the exact same variables, and I'm using the `dedupExchange`, so I'm not sure what it could be. It's happening both on initial mount and when variables change. Thanks!

**Comment by JoviDeCroock** [comment]
A bit light on the details 😅 are two queries running in parallel that stem from the same or two different components? What exchanges are involved? What policy are you using?

Quickly tried [here](https://codesandbox.io/s/romantic-perlman-83g3ux?file=/src/index.js) with the latest versions and only see 1 request upon initial mount.

![Image showing only 1 network call in relation to the above codesandbox](https://user-images.githubusercontent.com/17125876/181575447-bc4872b8-089d-40e0-9bef-920c2de02133.png)

Maybe it's something wit how React-Native or something else does microticks as we rely on the `fetchExchange` being deferred by a microtick during the synchronous cache-read. Or it has something to do with cache-and-network and suspense which is expected to have two calls.

**Reply by nandorojo** [reply]
> A bit light on the details 😅 

I know sorry lol. Just wanted to see if this is a known/common thing or if it's some problem with my setup. In this case, I'm actually seeing it happen in the browser on our website. I'm also not using suspense, but some of them are indeed using `cache-and-network`. I found the discussion around those two but figured it didn't apply since I'm not using suspense. 

Given that it sounds like an unintended thing, I'll dig deeper into making a reproduction and isolating the cause.

**Reply by kitten** [reply]
One of them should be cancelled as long as the fetch implementation supports the `AbortController`.

Generally, in the new version of React the effects now take longer than our micro tick to kick in in most cases, which means that you'll see two requests of which the first one should be cancelled (given the above being the case).

Generally we do not recommend `cache-and-network` as a policy by default. Adding the `requestPolicyExchange` is often the more solid choice as you then delay issuing new operations on demand to the network if a certain TTL has passed.

**Reply by nandorojo** [reply]
> One of them should be cancelled as long as the fetch implementation supports the `AbortController`.

Right, I’m on a new version of Chrome so I assume I have support for it. Is it possible it’s not aborting in time? I’m not familiar with how `AbortController` works.

> Generally, in the new version of React the effects now take longer than our micro tick to kick in in most cases

In case it’s relevant, I’m still using React 17. The double request is happening in production mode, and our backend is getting hit twice within a few milliseconds for the same call.

> Generally we do not recommend cache-and-network as a policy by default.

Yeah I don’t use it everywhere, but some calls (such as a user’s bookings or payment method) need to be up-to-date. It’s not the default policy though. I am using `requestPolicyExchange`, after the dedup exchange and before the refocus exchange.

**Reply by nandorojo** [reply]
I found the problem.

I was doing this:

```ts
authExchange({
  willAuthError: () => true,
  getAuth: async => {
    const token = await getToken()

    if (token) return { token }
    return null
  }
  // ...
})
```

I mistakenly interpreted `willAuthError` as a way to say, "let's call `getAuth()` before this request". To avoid ever sending a stale token, I figured I wanted to call `getAuth` before every request. Even though my `getToken()` function is async, under the hood it typically returns a cached token so it's instant. 

I failed to realize that calling `willAuthError` re-fires the request, resulting in these duplicates.

I removed this logic, instead opting for an implementation with `didAuthError`, and my duplicate requests went away.

Some sort of `shouldGetAuth()` might be useful that fires before the request without triggering a second one, but it isn't that necessary. `firebase/auth` doesn't provide a first-class method of getting the expiration time of a token, so I'd have to use a hack to add that functionality to `willAuthError()`. 

In any case, this is now answered.
